From 6f31b9fbd23c01c33a1b418d49884a157c5d0e48 Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 12:02:05 -0400
Subject: [ethernet] mlx4: Set skb->csum_level for encapsulated checksum

Message-id: <84616b8498f2718f6bd91f017500a916546f1af5.1410425016.git.avadai@redhat.com>
Patchwork-id: 93715
O-Subject: [Patch RHEL7.1 V1 189/197] mlx4: Set skb->csum_level for encapsulated checksum
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Tom Herbert <therbert@google.com>

Set skb->csum_level instead of skb->encapsulation when indicating
CHECKSUM_UNNECESSARY for an encapsulated checksum.

Signed-off-by: Tom Herbert <therbert@google.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit 9ca8600ec1e2ea209b20a38d618ab2c00e86da28)
Signed-off-by: Amir Vadai <avadai@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/mellanox/mlx4/en_rx.c b/drivers/net/ethernet/mellanox/mlx4/en_rx.c
index 1f394c4..6892512 100644
--- a/drivers/net/ethernet/mellanox/mlx4/en_rx.c
+++ b/drivers/net/ethernet/mellanox/mlx4/en_rx.c
@@ -768,7 +768,7 @@ int mlx4_en_process_rx_cq(struct net_device *dev, struct mlx4_en_cq *cq, int bud
      gro_skb->ip_summed = CHECKSUM_UNNECESSARY;
 
      if (l2_tunnel)
-      gro_skb->encapsulation = 1;
+      gro_skb->csum_level = 1;
      if ((cqe->vlan_my_qpn &
          cpu_to_be32(MLX4_CQE_VLAN_PRESENT_MASK)) &&
          (dev->features & NETIF_F_HW_VLAN_CTAG_RX)) {
@@ -822,8 +822,8 @@ int mlx4_en_process_rx_cq(struct net_device *dev, struct mlx4_en_cq *cq, int bud
   skb->protocol = eth_type_trans(skb, dev);
   skb_record_rx_queue(skb, cq->ring);
 
-  if (l2_tunnel)
-   skb->encapsulation = 1;
+  if (l2_tunnel && ip_summed == CHECKSUM_UNNECESSARY)
+   skb->csum_level = 1;
 
   if (dev->features & NETIF_F_RXHASH)
    skb_set_hash(skb,
-- 
1.7.1